requests_kerberos/kerberos_.py

```diff
@@ -161,14 +162,44 @@ def _get_channel_bindings_application_data(response):
 
     return application_data
 
+
+# Return the hostname as it will be canonicalized by
+# krb5_sname_to_principal.  We can't simply use socket.getfqdn()
+# because it explicitly prefers results containing periods and
+# krb5_sname_to_principal doesn't care.
+#
+# This is necessary to support use cases where multiple services
+# are hosted on the same machine under multiple CNAMEs. It won't do
+# any good to use a SPN that mentions one of the CNAMES, we need to ask for the real machine.
+#
+# Maybe we should use socket.getfqdn here, but that doesn't use AI_CANONNAME
+# but rather chooses the first entry in gethostbyaddr that contains a dot, which seems
+# a trifle strange: https://github.com/python/cpython/blob/master/Lib/socket.py#L681
+#
+# https://github.com/krb5/krb5/blob/d406afa363554097ac48646a29249c04f498c88e/src/util/k5test.py#L505-L520
+# https://chromium.googlesource.com/chromium/src/+/lkgr/net/http/http_auth_handler_negotiate.cc#150
+def _get_default_kerb_host(hostname):
+    try:
+        ais = socket.getaddrinfo(hostname, None, 0, 0, 0, socket.AI_CANONNAME)
+    except socket.gaierror:
+        ais = []
+    
+    if not ais:
+        return hostname.lower()
+
+    _family, _socktype, _proto, canonname, _sockaddr = ais[0]
```

Comment:
Based on how we intend to use this, I'm wondering if we should filter out anything that has a family which isn't AF_INET or AF_INET6, as well as anything which has a socktype which isn't SOCK_STREAM, as well as anything which doesn't have a photo of IPPROTO_TCP.

Do you think that's paranoid? 

I'm not certain whether or not cannoname is expected to be the same for all results in the ais list. It is for everything I tried, but that's different from verification of the spec.

Reply:
I'm not sure that we should do extra filtering. By doing so I think we're just as likely to break some obscure deployment as we are to fix any problems. What if someone is still using IPX networking, or running requests-kerberos in the year 2045 over IPv7?

I couldn't find any definitive answer about this, but it does seem to me that canonname should be the same across all entries. It's a property of the host, not the particular socket/address that is used to communicate with it.